### ISSUE TITLE
Setting the car current charge rate throws an exception

### ISSUE THREAD

**Issue by i-am-shodan** [alandtse/tesla#971]
### Is there an existing issue for this?

- [X] I have searched both the existing open issues & recently closed issues and did not find a duplicate of this issue.

### I have read about the Fleet API and understand I may need to use it

- [X] I understand issues relating to read only commands will be auto closed if not using the Fleet API.

### Version of the Tesla component

V3.21.0

### Version of the Tesla car software

2024.8.9

### Model

Model Y

### Current Behavior

Up until the most recent version I could happily set the charge current current using

```
service: number.set_value
metadata: {}
data:
    value: "14"
  target:
    entity_id: number.lightning_charging_amps
```

now I get a TeslaException thrown

### Expected Behavior

Charge rate is set, exception isn’t throw 

### Debug logs

```true
logger:%20homeassistant.components.automation.solar_management%0ASource%3A%20components/automation/__init__.py:723%0Aintegration:%20Automation%20(documentation,%20issues)%0AFirst%20occurred:%2025%20May%202024%20at%2015:46:34%20(5%20occurrences)%0ALast%20logged:%2013:15:20%0A%0AWhile%20executing%20automation%20automation.solar_management%0ATraceback%20(most%20recent%20call%20last):%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/components/automation/__init__.py%22,%20line%20723,%20in%20async_trigger%0A%20%20%20%20return%20await%20self.action_script.async_run(%0A%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%201731,%20in%20async_run%0A%20%20%20%20return%20await%20asyncio.shield(create_eager_task(run.async_run()))%0A%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20451,%20in%20async_run%0A%20%20%20%20await%20self._async_step(log_exceptions=False)%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20503,%20in%20_async_step%0A%20%20%20%20self._handle_exception(%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20533,%20in%20_handle_exception%0A%20%20%20%20raise%20exception%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20501,%20in%20_async_step%0A%20%20%20%20await%20getattr(self,%20handler)()%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%201049,%20in%20_async_if_step%0A%20%20%20%20await%20self._async_run_script(if_data%5B%22if_else%22%5D)%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%201210,%20in%20_async_run_script%0A%20%20%20%20result%20=%20await%20self._async_run_long_action(%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20699,%20in%20_async_run_long_action%0A%20%20%20%20return%20await%20long_task%0A%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%201731,%20in%20async_run%0A%20%20%20%20return%20await%20asyncio.shield(create_eager_task(run.async_run()))%0A%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20451,%20in%20async_run%0A%20%20%20%20await%20self._async_step(log_exceptions=False)%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20503,%20in%20_async_step%0A%20%20%20%20self._handle_exception(%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20533,%20in%20_handle_exception%0A%20%20%20%20raise%20exception%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20501,%20in%20_async_step%0A%20%20%20%20await%20getattr(self,%20handler)()%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%201049,%20in%20_async_if_step%0A%20%20%20%20await%20self._async_run_script(if_data%5B%22if_else%22%5D)%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%201210,%20in%20_async_run_script%0A%20%20%20%20result%20=%20await%20self._async_run_long_action(%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20699,%20in%20_async_run_long_action%0A%20%20%20%20return%20await%20long_task%0A%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%201731,%20in%20async_run%0A%20%20%20%20return%20await%20asyncio.shield(create_eager_task(run.async_run()))%0A%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20451,%20in%20async_run%0A%20%20%20%20await%20self._async_step(log_exceptions=False)%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20503,%20in%20_async_step%0A%20%20%20%20self._handle_exception(%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20533,%20in%20_handle_exception%0A%20%20%20%20raise%20exception%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20501,%20in%20_async_step%0A%20%20%20%20await%20getattr(self,%20handler)()%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%201043,%20in%20_async_if_step%0A%20%20%20%20await%20self._async_run_script(if_data%5B%22if_then%22%5D)%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%201210,%20in%20_async_run_script%0A%20%20%20%20result%20=%20await%20self._async_run_long_action(%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20699,%20in%20_async_run_long_action%0A%20%20%20%20return%20await%20long_task%0A%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%201731,%20in%20async_run%0A%20%20%20%20return%20await%20asyncio.shield(create_eager_task(run.async_run()))%0A%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20451,%20in%20async_run%0A%20%20%20%20await%20self._async_step(log_exceptions=False)%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20503,%20in%20_async_step%0A%20%20%20%20self._handle_exception(%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20533,%20in%20_handle_exception%0A%20%20%20%20raise%20exception%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20501,%20in%20_async_step%0A%20%20%20%20await%20getattr(self,%20handler)()%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20736,%20in%20_async_call_service_step%0A%20%20%20%20response_data%20=%20await%20self._async_run_long_action(%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/script.py%22,%20line%20699,%20in%20_async_run_long_action%0A%20%20%20%20return%20await%20long_task%0A%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/core.py%22,%20line%202738,%20in%20async_call%0A%20%20%20%20response_data%20=%20await%20coro%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/core.py%22,%20line%202779,%20in%20_execute_service%0A%20%20%20%20return%20await%20target(service_call)%0A%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/service.py%22,%20line%20975,%20in%20entity_service_call%0A%20%20%20%20single_response%20=%20await%20_handle_entity_call(%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/helpers/service.py%22,%20line%201047,%20in%20_handle_entity_call%0A%20%20%20%20result%20=%20await%20task%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/src/homeassistant/homeassistant/components/number/__init__.py%22,%20line%20116,%20in%20async_set_value%0A%20%20%20%20await%20entity.async_set_native_value(native_value)%0A%20%20File%20%22/config/custom_components/tesla_custom/number.py%22,%20line%2091,%20in%20async_set_native_value%0A%20%20%20%20await%20self._car.set_charging_amps(value)%0A%20%20File%20%22/usr/local/lib/python3.12/site-packages/teslajsonpy/car.py%22,%20line%20897,%20in%20set_charging_amps%0A%20%20%20%20data%20=%20await%20self._send_command(%0A%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/local/lib/python3.12/site-packages/teslajsonpy/car.py%22,%20line%20775,%20in%20_send_command%0A%20%20%20%20raise%20ex%0A%20%20File%20%22/usr/local/lib/python3.12/site-packages/teslajsonpy/car.py%22,%20line%20760,%20in%20_send_command%0A%20%20%20%20data%20=%20await%20self._controller.api(%0A%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/local/lib/python3.12/site-packages/teslajsonpy/controller.py%22,%20line%201367,%20in%20api%0A%20%20%20%20response%20=%20await%20self.__post_with_retries(%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/local/lib/python3.12/site-packages/tenacity/_asyncio.py%22,%20line%2088,%20in%20async_wrapped%0A%20%20%20%20return%20await%20fn(*args,%20**kwargs)%0A%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/local/lib/python3.12/site-packages/tenacity/_asyncio.py%22,%20line%2047,%20in%20__call__%0A%20%20%20%20do%20=%20self.iter(retry_state=retry_state)%0A%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/local/lib/python3.12/site-packages/tenacity/__init__.py%22,%20line%20325,%20in%20iter%0A%20%20%20%20raise%20retry_exc.reraise()%0A%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/local/lib/python3.12/site-packages/tenacity/__init__.py%22,%20line%20158,%20in%20reraise%0A%20%20%20%20raise%20self.last_attempt.result()%0A%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/local/lib/python3.12/concurrent/futures/_base.py%22,%20line%20449,%20in%20result%0A%20%20%20%20return%20self.__get_result()%0A%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/local/lib/python3.12/concurrent/futures/_base.py%22,%20line%20401,%20in%20__get_result%0A%20%20%20%20raise%20self._exception%0A%20%20File%20%22/usr/local/lib/python3.12/site-packages/tenacity/_asyncio.py%22,%20line%2050,%20in%20__call__%0A%20%20%20%20result%20=%20await%20fn(*args,%20**kwargs)%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/local/lib/python3.12/site-packages/teslajsonpy/controller.py%22,%20line%201390,%20in%20__post_with_retries%0A%20%20%20%20return%20await%20self.__connection.post(command,%20method=method,%20data=data,%20url=url)%0A%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/local/lib/python3.12/site-packages/teslajsonpy/connection.py%22,%20line%20166,%20in%20post%0A%20%20%20%20return%20await%20self.__open(url,%20method=method,%20headers=self.head,%20data=data)%0A%20%20%20%20%20%20%20%20%20%20%20%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%5E%0A%20%20File%20%22/usr/local/lib/python3.12/site-packages/teslajsonpy/connection.py%22,%20line%20218,%20in%20__open%0A%20%20%20%20raise%20TeslaException(resp.status_code)%0Ateslajsonpy.exceptions.TeslaException%0A
```

**Comment by i-am-shodan** [comment]
If I change the slider in the entities page I also get an error so it's unrelated to the yaml above

**Comment by i-am-shodan** [comment]
I can see I am getting updates so my token hasn't expired

**Comment by skipishere** [comment]
Are you using the Fleet API?
Do any other commands work or raise exceptions?

**Comment by i-am-shodan** [comment]
I don't even know what the fleet api is. I just set it up from the integration screen.

**Comment by i-am-shodan** [comment]
Looks like changing the charge limit doesn't work either

**Comment by skipishere** [comment]
If you don't know what the fleet api is (which kind of implies you just ticked the box 
> I have read about the Fleet API and understand I may need to use it
> I understand issues relating to read only commands will be auto closed if not using the Fleet API.
without reading it... Please take a look at https://github.com/alandtse/tesla?tab=readme-ov-file#tesla-fleet-api-proxy as it explains what it is and why you need to use it (as it appears you can't use the older and unofficial API)

**Comment by i-am-shodan** [comment]
Thanks for getting back to me. I'm sure setting up the API proxy will fix the issue. Feel free to close the issue and I'll report back if I see the same issue with the new API.

If I may, I'd like to give comment on the upgrade path.

I'm no expert on the Tesla API. I've no idea if before I was using the owner API or not. The docs say you'll see errors in the logs, this isn't very specific especially as from my perspective it was working fine before.

I looked in the configuration for the integration and I couldn't see mention of either the OwnerAPI or the FleetAPI. Perhaps you could make it clear somewhere what the API in use is? Better yet disable the update ability to make it clear this isn't supported. The perfect solution would be for the issue to appear in the Repair section of HA.

From my perspective I've a time poor use of HACS. I smash upgrade and hope for the best. It would be great if the upgrade path was really clear that clicking upgrade in HACS isn't enough to get the integration working again. I clearly missed this. Maybe even a big red box at the top of the Readme.md.

Catching the exception and logging a message saying you need to migrate to this new API also seems like a good idea to me. All these things immediate push the work on me to fix.

All that being said this integration is really important to me. It literally saves me a ton of money by enabling me to trickle charge my car when the sun is shining. Thanks for all the hard work keeping it running.

**Comment by skipishere** [comment]
Sadly Tesla have made this very hard, given the way they have made the Fleet API (and that it's made with fleets in mind, not normal consumers) and that some older cars can't move over to it. Some (like mine) are expected to be forced onto the Fleet API (as of start of the month I think from the last time they moved the deadlines) but still work for now whereas others got forced over at the start of the year. Ideally if we could capture the exception and make it clearer that would be great (we'd welcome a PR that did that, I'm no python expert and just try and help out with issue triaging)

Also (just to warn you) the Fleet API isn't perfect as it is, the need for us to create a dev account with Tesla and that (for now) they aren't charging for use but 'might' do in the future and there are rate limits (which you can easily hit). Hopefully Tesla might listen to the community feedback, but so far it's not been great.

I'll close this for now, but if you do get the Fleet API working for you and this issue still persists, feel free to reopen it.

**Comment by Peter919pan** [comment]
I wanted to express my sincere gratitude to alandtse and skipishere for their contributions! Thanks to their work, I've been able to significantly reduce my electricity costs, similar to Shodan.

I followed the instructions to create a developer account on https://developer.tesla.com/ and successfully got approval for my application. However, I'm currently stuck on obtaining the Client ID and Client Secret.

Here's the issue:

I received an email confirming my application's approval, but there's nothing showing in my Application Dashboard.
The Application Dashboard displays "No Application Yet" and a "New Application" button.
I hope this message doesn't bother alandtse or skipishere. I'm simply wondering if anyone else has encountered the same problem.

Once again, a massive thank you to alandtse, skipishere, and wish Shodan or similar people to resolve this soon!

![Screenshot 2024-05-26 at 15 56 42](https://github.com/alandtse/tesla/assets/170870808/624b6fa4-eb28-4751-8bb4-8503bce69e6e)
